File: Topic1_Arrays/Day2207/exercise_2.cpp
// Use DFS to count the number of islands
// Time complexity : O(n^2)
// Space complexity : O(n) (The number of recusive function calls in the call stack)
class Solution {
public:
    int dx[4] = {-1, 0, 0, 1};
    int dy[4] = {0, -1, 1, 0};
    void loang(int i, int j, vector<vector<char>>& grid) {
        grid[i][j] = '0';
        for (int k = 0; k < 4; k++) {
            int i1 = i + dx[k], j1 = j + dy[k];
            if (i1 >= 0 && i1 < grid.size() && j1 >= 0 && j1 < grid[i].size() &&
                grid[i1][j1] == '1') {
                loang(i1, j1, grid);
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == '1') {
                    loang(i, j, grid);
                    res++;
                }
            }
        }
        return res;
    }
};

File: Topic1_Arrays/Day2207/Exercise2.cs
//https://leetcode.com/problems/number-of-islands/description/

/*
+ To use recursion technique
+ I will loop every single element in 2d arrays
+ If current value is '1' => not visited => backtracking that position and plus 1 value of result
+ When backtracking current position, mark its neightbor values as '2' (islands but visited)

Space complexity: O(n^2)
Time complexity: O(1)
*/

namespace Day2207
{
    public class Exercise2
    {
        public int NumIslands(char[][] grid)
        {
            int row = grid.Length;
            int col = grid[0].Length;

            int result = 0;

            for (int i = 0; i < row; ++i)
            {
                for (int j = 0; j < col; ++j)
                {
                    if (grid[i][j] == '1')
                    {
                        ++result;
                        Backtrack(grid, i, j, row, col);
                    }
                }
            }

            return result;
        }

        private void Backtrack(char[][] grid, int i, int j, int m, int n)
        {
            if (i < 0 || i == m || j < 0 || j == n || grid[i][j] != '1')
                return;

            grid[i][j] = '2'; // visited

            Backtrack(grid, i - 1, j, m, n);
            Backtrack(grid, i, j - 1, m, n);
            Backtrack(grid, i + 1, j, m, n);
            Backtrack(grid, i, j + 1, m, n);
        }
    }
}